### PR TITLE
Playlist management

### PR DESCRIPTION
**Pull request by ShekharDhangar**
Added Playlist Management.  Added Single video Page.
**Number of lines added are not 1002,
Around 500 lines are of data only so no need to check data objects.
And Some files are already review in previos PR like Authentication Context and related Auth files.**
In this PR, Check changes related to Playlist and Single Video Page.
Also, chips have been added through categories.
Please do give your reviews.
Files to check:
[App](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-3d74dddefb6e35fbffe3c76ec0712d5c416352d9449e2fcc8210a9dee57dff67)
[PlaylistMenuComponent](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-897943f0554f4aa46767a8ddc726a306d4a65fef905c9a948a1a6bdf4bae4676)
[PlaylistPageCard](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-87a62b13feeacaaaee4adfd808e41c40e9ad473320af9131e965456d3a6d9f76)
[SingleVideoPlayer](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-a522ae646cbf3c6e3f9593192b1c246d325509fdd4a52741f99ad1a41cec07f8)
[PlaylistFormBox](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-a28e48e458d7aa1df9419df644bc4906e4c819c98088d215f38df1925f7e78d5)
[MenuBox](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-18a044023f91f3b38b605c6a968482d72459781302f5830f67a3d5c7d513a6c5)-Abstracted out from video component
[VideoCard](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-f891f8c9cbca251ffa27ed4933a7b8356c4249fd31b61701e0fc041bd135e126)
[HomePage](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-bbf51f89c05ba45861a2e102d3dedb2c01b263f6c77edc5d4dffed234b8181f7)
[playlistPage](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-9952380ab81e951f8dd55f713ad46e7f305719f93adb4215f7285596ae496f46)
[PlaylistVideoList](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-39b60e07ac4535ecc4ead8f47a2028823e643de22c16f8053833363207c68dbd)
[SingleVideoPage](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-102a476a0ec39b6fe7c2ec13c1e75feebee248508cf636f1ebf32837cf3ac172)
[Reducer](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-2153ec3493046853fe510118e935cd85510bae579392cb294f33ee9b73ec1d23)
[InitialCalls](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-6e4e202fce30d69e5d50ff41e7cf6ccdee19cd2559564588f070f9e76cdc849c)
[PlaylistServerCallls](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-808904f609ff960689005ef892c65957e2270d18ab898ae0c9c7dcb11963a67b)
[operations](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-2a37d7792201203ddd97f5b0061aced07278e81faffbecfd998ec163a249031d)
[Utilscalls](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-f8804721d96694d0e905d77e6f7d1be1e9975d5c7ca2e637e0081ee932d3da3e)

[playlistPageCard-CSS](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-87a62b13feeacaaaee4adfd808e41c40e9ad473320af9131e965456d3a6d9f76)-CSS
[SinglePageVideo-CSS](https://github.com/ShekharDhangar/shed-video-store/pull/4/files?file-filters%5B%5D=.css&file-filters%5B%5D=.js&file-filters%5B%5D=.jsx&show-viewed-files=true&show-deleted-files=false#diff-d58ace358f851bbd8ce701106c420ef9d4a0202f2e366f6c43cc61857b98d2f9)

### CODE REVIEW

File: src/components/VideoCard/AddPlaylistBox/AddPlaylistBox.jsx [hunk 1]
```diff
@@ -1,10 +1,46 @@
...
+    title: "",
+    description: "",
+  });
+
+  const newPlaylistHandler = (e) => {
     e.preventDefault();
-  }
+    if (userState.id) {
+      if (playlists.some((item) => item.title === inputState.title)) {
+      } else {
```
Comment (partha8) [inline src/components/VideoCard/AddPlaylistBox/AddPlaylistBox.jsx]: There's an if statement without any logic in it.

File: src/components/VideoCard/AddPlaylistBox/AddPlaylistBox.jsx [hunk 2]
```diff
@@ -1,10 +1,46 @@
...
+        addPlaylistHandler(inputState, userState?.id, dispatch);
```
Comment (partha8) [inline src/components/VideoCard/AddPlaylistBox/AddPlaylistBox.jsx]: suggestion: you can clear out the input field after you create a new playlist.

File: src/reducer/videoReducer.js [hunk 3]
```diff
@@ -1,14 +1,34 @@
...
+    // case "REMOVE_VIDEO_TO_PLAYLIST": {
+    //     return {
+    //        ...state,
+    //        playlists: state.playlists.reduce((acc, curr) => {
+    //          curr._id === payload.playlistId
+    //            ? (acc = [...acc, payload.addedPlaylistVideo])
+    //            : (acc = [...acc, curr]);
+    //        }, []),
+    //      };
+    //    }
```
Comment (partha8) [inline src/reducer/videoReducer.js]: remove comments

**Review by partha8 — APPROVED** [review]
Overall good work, spent a good amount of time going through the code and it was readable.
Good use of shuffle algorithm, last night I was looking at it as well.
Good UI as well. :rocket:

edit:
Add an alert when a video is successfully added to the playlist, I was lost for a second after clicking on the checkbox.

File: src/components/PlaylistPageCard/PlaylistPageCard.css [hunk 4]
```diff
@@ -0,0 +1,23 @@
+.card-overlay-content{
+    color: var(--clr-white);
+    justify-content: center;
+    align-items: flex-end;
+}
+.content{
```
Comment (hsnice16) [inline src/components/PlaylistPageCard/PlaylistPageCard.css:6]: I think it is nice to maintain at least a line space between two declarations.

File: src/components/PlaylistMenuBtn/PlaylistMenuBtn.jsx [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import { useVideoContext } from "../../context/videoContext";
+import {FaTrash} from "../../icons/icons";
+import { useParams } from "react-router-dom";
+import { removeVideoFromPlaylistCall } from "../../utils/utils";
+import { useAuthContext } from "../../context/AuthContext";
+export function PlaylistMenuBtn({videoId}){
+    const {dispatch}=useVideoContext();
+    const {userState}=useAuthContext();
+    const {playlistId}=useParams();
+    return (
```
Comment (hsnice16) [inline src/components/PlaylistMenuBtn/PlaylistMenuBtn.jsx:10]: You can give a line space here. Before `return`.

File: src/context/videoContext.js [hunk 6]
```diff
@@ -3,7 +3,7 @@ import { videoReducer } from "../reducer/videoReducer";
 const VideoContext = createContext();
```
Comment (hsnice16) [inline src/context/videoContext.js]: File naming suggestions,
I think you should not use the naming `videoContext` here, as in javascript usually, we name a function like this. And in this file nothing exist with the name `videoContext`.
I think, you can name the way Tanay had suggested, `video-context`, or anything else you find good other than this.

File: src/App.js [hunk 7]
```diff
@@ -1,23 +1,30 @@
...
 import { RoutesPath } from "./Routes";
-import { getVideos } from "./serverCalls/getVideos";
-import {useVideoContext} from "./context/videoContext";
+import { getVideos, getCategories } from "./serverCalls/initialCalls";
+import { useVideoContext } from "./context/videoContext";
 
 function App() {
-  const {dispatch} = useVideoContext();
+  const { dispatch } = useVideoContext();
   useEffect(() => {
```
Comment (hsnice16) [inline src/App.js]: Why not make this API call inside the `videoContext`? 
It seems like there is nothing that is specific to the `App` file and you are using it. So why not make this API call in the context?

**Review by hsnice16 — APPROVED** [review]
Nice work. Have added a few suggestions.

**Comment by hsnice16** [comment]
Some suggestions on the UI, 
In this modal, it is not closing on clicking the backdrop area (means clicking outside the modal box). I think, it is a common behaviour you can check on some other sites for inspiration.

image

On single video page, the video this has been clicked should not be shown in the side list of videos. 
image

File: src/components/VideoCard/AddPlaylistBox/AddPlaylistBox.jsx (hunk 2 above)
Comment (ShekharDhangar) [reply src/components/VideoCard/AddPlaylistBox/AddPlaylistBox.jsx]: Made the change

File: src/reducer/videoReducer.js (hunk 3 above)
Comment (ShekharDhangar) [reply src/reducer/videoReducer.js]: Removed

**Comment by ShekharDhangar** [comment]
> Overall good work, spent a good amount of time going through the code and it was readable. Good use of shuffle algorithm, last night I was looking at it as well. Good UI as well. 🚀
> 
> edit: Add an alert when a video is successfully added to the playlist, I was lost for a second after clicking on the checkbox.

Actually, I tried but was facing some issue so I'm going to add toast,alerts at the end of submission if got time.

File: src/components/PlaylistPageCard/PlaylistPageCard.css (hunk 4 above)
Comment (ShekharDhangar) [reply src/components/PlaylistPageCard/PlaylistPageCard.css:6]: actually its my code formatter, so it happens everytime

File: src/components/PlaylistMenuBtn/PlaylistMenuBtn.jsx (hunk 5 above)
Comment (ShekharDhangar) [reply src/components/PlaylistMenuBtn/PlaylistMenuBtn.jsx:10]: done

File: src/context/videoContext.js (hunk 6 above)
Comment (ShekharDhangar) [reply src/context/videoContext.js]: made the change

File: src/App.js (hunk 7 above)
Comment (ShekharDhangar) [reply src/App.js]: Made the changes, done

**Comment by ShekharDhangar** [comment]
> Some suggestions on the UI, In this modal, it is not closing on clicking the backdrop area (means clicking outside the modal box). I think, it is a common behaviour you can check on some other sites for inspiration.
> 
> image
> 
> On single video page, the video this has been clicked should not be shown in the side list of videos. image

made both changes

**Pull request merged**